From 81cc5406dcb90ae019c919d3c5247c3f7eec23ef Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Aur=C3=A9lien=20Brooke?= <aurelien@bahiasoft.fr>
Date: Fri, 14 Aug 2026 13:14:39 +0200
Subject: [PATCH] rhi: gl: support unsigned integer uniform arrays

QRhiGles2 emulates uniform buffers with individual uniforms. The
unsigned integer upload paths always set a single element, unlike their
floating-point and signed integer counterparts. This leaves all but the
first element of uint, uvec2, uvec3, and uvec4 arrays unset.

Upload every reflected array element. Remove the std140 padding from
uint, uvec2, and uvec3 arrays before passing them to OpenGL; uvec4
arrays can be uploaded directly. Validate the complete array range
before reading from the buffer.

[ChangeLog][RHI] The OpenGL backend now supports arrays of uint, uvec2,
uvec3, and uvec4 members in uniform buffers.

Change-Id: I2a226707e8874a2b464a1987006b59159f27447d
Reviewed-by: Laszlo Agocs <laszlo.agocs@qt.io>
---
 src/gui/rhi/qrhigles2.cpp | 49 ++++++++++++++++++++++++++++++++-------
 src/gui/rhi/qrhigles2_p.h |  1 +
 2 files changed, 42 insertions(+), 8 deletions(-)

diff --git a/src/gui/rhi/qrhigles2.cpp b/src/gui/rhi/qrhigles2.cpp
index ac6692a4d12..ef6ecff96e3 100644
--- a/src/gui/rhi/qrhigles2.cpp
+++ b/src/gui/rhi/qrhigles2.cpp
@@ -4443,18 +4443,22 @@ static inline qint64 qrhi_std140_read_size(QShaderDescription::VariableType type
     switch (type) {
     case QShaderDescription::Float:
     case QShaderDescription::Int:
+    case QShaderDescription::Uint:
         elemSize = 4;
         break;
     case QShaderDescription::Vec2:
     case QShaderDescription::Int2:
+    case QShaderDescription::Uint2:
         elemSize = 8;
         break;
     case QShaderDescription::Vec3:
     case QShaderDescription::Int3:
+    case QShaderDescription::Uint3:
         elemSize = 12;
         break;
     case QShaderDescription::Vec4:
     case QShaderDescription::Int4:
+    case QShaderDescription::Uint4:
         elemSize = 16;
         break;
     case QShaderDescription::Mat2:
@@ -4474,13 +4478,9 @@ static inline qint64 qrhi_std140_read_size(QShaderDescription::VariableType type
-    case QShaderDescription::Uint:
     case QShaderDescription::Bool:
         return 4;
-    case QShaderDescription::Uint2:
     case QShaderDescription::Bool2:
         return 8;
-    case QShaderDescription::Uint3:
     case QShaderDescription::Bool3:
         return 12;
-    case QShaderDescription::Uint4:
     case QShaderDescription::Bool4:
         return 16;
     default:
@@ -4600,13 +4601,18 @@ void QRhiGles2::bindShaderResources(QGles2CommandBuffer *cbD,
                             && uniform.type != QShaderDescription::Int2
                             && uniform.type != QShaderDescription::Int3
                             && uniform.type != QShaderDescription::Int4
+                            && uniform.type != QShaderDescription::Uint
+                            && uniform.type != QShaderDescription::Uint2
+                            && uniform.type != QShaderDescription::Uint3
+                            && uniform.type != QShaderDescription::Uint4
                             && uniform.type != QShaderDescription::Mat2
                             && uniform.type != QShaderDescription::Mat3
                             && uniform.type != QShaderDescription::Mat4)
                     {
                         qWarning("Uniform with buffer binding %d, buffer offset %d, type %d is an array, "
                                  "but arrays are only supported for float, vec2, vec3, vec4, int, "
-                                 "ivec2, ivec3, ivec4, mat2, mat3 and mat4. "
+                                 "ivec2, ivec3, ivec4, uint, uvec2, uvec3, uvec4, mat2, mat3 "
+                                 "and mat4. "
                                  "Only the first element will be set.",
                                  uniform.binding, uniform.offset, uniform.type);
                     }
@@ -4802,16 +4808,43 @@ void QRhiGles2::bindShaderResources(QGles2CommandBuffer *cbD,
                         f->glUniform4iv(uniform.glslLocation, qMax(1, uniform.arrayDim), reinterpret_cast<const qint32 *>(src));
                         break;
                     case QShaderDescription::Uint:
-                        f->glUniform1ui(uniform.glslLocation, *reinterpret_cast<const quint32 *>(src));
+                    {
+                        const int elemCount = uniform.arrayDim;
+                        if (elemCount < 1) {
+                            f->glUniform1ui(uniform.glslLocation, *reinterpret_cast<const quint32 *>(src));
+                        } else {
+                            m_scratch.packedArray.resize(elemCount);
+                            qrhi_std140_to_packed(&m_scratch.packedArray.data()->u, 1, elemCount, src);
+                            f->glUniform1uiv(uniform.glslLocation, elemCount, &m_scratch.packedArray.constData()->u);
+                        }
+                    }
                         break;
                     case QShaderDescription::Uint2:
-                        f->glUniform2uiv(uniform.glslLocation, 1, reinterpret_cast<const quint32 *>(src));
+                    {
+                        const int elemCount = uniform.arrayDim;
+                        if (elemCount < 1) {
+                            f->glUniform2uiv(uniform.glslLocation, 1, reinterpret_cast<const quint32 *>(src));
+                        } else {
+                            m_scratch.packedArray.resize(elemCount * 2);
+                            qrhi_std140_to_packed(&m_scratch.packedArray.data()->u, 2, elemCount, src);
+                            f->glUniform2uiv(uniform.glslLocation, elemCount, &m_scratch.packedArray.constData()->u);
+                        }
+                    }
                         break;
                     case QShaderDescription::Uint3:
-                        f->glUniform3uiv(uniform.glslLocation, 1, reinterpret_cast<const quint32 *>(src));
+                    {
+                        const int elemCount = uniform.arrayDim;
+                        if (elemCount < 1) {
+                            f->glUniform3uiv(uniform.glslLocation, 1, reinterpret_cast<const quint32 *>(src));
+                        } else {
+                            m_scratch.packedArray.resize(elemCount * 3);
+                            qrhi_std140_to_packed(&m_scratch.packedArray.data()->u, 3, elemCount, src);
+                            f->glUniform3uiv(uniform.glslLocation, elemCount, &m_scratch.packedArray.constData()->u);
+                        }
+                    }
                         break;
                     case QShaderDescription::Uint4:
-                        f->glUniform4uiv(uniform.glslLocation, 1, reinterpret_cast<const quint32 *>(src));
+                        f->glUniform4uiv(uniform.glslLocation, qMax(1, uniform.arrayDim), reinterpret_cast<const quint32 *>(src));
                         break;
                     case QShaderDescription::Bool: // a glsl bool is 4 bytes, like (u)int
                         f->glUniform1i(uniform.glslLocation, *reinterpret_cast<const qint32 *>(src));
diff --git a/src/gui/rhi/qrhigles2_p.h b/src/gui/rhi/qrhigles2_p.h
index e919374e1a4..ef82717dc46 100644
--- a/src/gui/rhi/qrhigles2_p.h
+++ b/src/gui/rhi/qrhigles2_p.h
@@ -1189,6 +1189,7 @@ public:
         union data32_t {
             float f;
             qint32 i;
+            quint32 u;
         };
         QVarLengthArray<data32_t, 128> packedArray;
         struct SeparateTexture {
-- 
2.54.0.windows.1

